src/base/Group.h

```diff
@@ -0,0 +1,110 @@
+#pragma once
+
+#include <algorithm>
+#include <set>
+#include <boost/iterator/indirect_iterator.hpp>
+
+namespace NuTo
+{
+namespace Groups
+{
+
+template <typename T>
+class Group : private std::set<T*>
+{
```

Comment:
> [weak_ptr](http://en.cppreference.com/w/cpp/memory/weak_ptr) is a smart pointer that holds a non-owning ("weak") reference to an object that is managed by std::shared_ptr. It must be converted to std::shared_ptr in order to access the referenced object.

So `weak_ptr` would require that each object is allocated as a shared_ptr. This is not the way to go.

What we are actually looking for is 
`std::InsertContainerHere<`[`std::reference_wrapper`](http://en.cppreference.com/w/cpp/utility/functional/reference_wrapper)`<T>>`. This would work without the indirect iterator, but the syntax would include `get()`:
~~~c++
for (auto entity : group)
    entity.get()->MemberFunction();
~~~

---
The whole out-of-scope topic did not occur because our _god class_ `structure` owned every object. IMO we want to eliminate that and have to take care of the scope. Technically, you can mess up whenever you have a reference or pointer member.
~~~c++
NuTo::NewmarkDirect CreateNewmark(...)
{
    NuTo::Structure s;
    s.Import(...)
    // ...
    return NuTo::NewmarkDirect(&s); 
} // Boom. s out of scope.
~~~

---
> Users don't know if it allows duplicates or if it is sorted or not.

I got your hint, I will add documentation. But what should the requirements be like? IMO: 
- no duplicates
- not sorted. Meaning that you should not rely on any sorting. Internally, some sorting might occur for performance reasons.